submariner-operator/broker-multi-tenancy.md

```diff
@@ -0,0 +1,321 @@
+1. Setup RBAC
+    1. Ensure broker admin SA `submariner-k8s-broker-admin` exists in broker NS.
+    1. Ensure broker admin role exists in broker NS `submariner-k8s-broker-admin`.
+    *Ensure it has the expected permissions to mutate submariner broker-related CRs.
+    Fail CLI if unable to mutate permissions to the expected set.*
+    1. Ensure role binding exists in broker NS (SA + role)
+    1. Backward compatibility - create legacy RBAC:
+        1. *Skip this in case `tenants.submariner.io` CRD exists - this means we
+        are in Axon mode, and we do not need to support legacy RBAC (and also cannot
+        create a cluster role).*
+        1. Create broker client service account in NS if not exist (`submariner-k8s-broker-client`)
+        1. Create broker cluster cluster-role in broker NS (`submariner-k8s-broker-cluster`)
+        1. Create role binding
+    1. Wait for broker admin SA secret token to be ready
+1. *Deploy operator (deleted, will not be carried on anymore)*
+1. Deploy broker
+    - Create or replace broker object in broker NS with spec parameters
+    from `subctl` CLI
+1. Write broker info local file (`broker-info.subm`)
+    1. Backup `.subm` file if already exists
+    1. Construct `.subm` file state
+        1. Read client secret token from `submariner-k8s-broker-admin` SA
+        1. Read ipsec PSK from old `.subm` file or generate a new PSK.
+        1. Read broker URL from current k8s context current URL in use
+        1. *Set `namespace` field to broker namespace.*
+        1. Update `.subm` file state from CLI args: components, service discovery,
+        custom domains.
+        1. Validate existing global networks
+        1. Create globalnet config map
+        1. Write `.subm` file to disk
+
+## Alternatives
+
+- Running multiple brokers (e.g., dedicated clusters or kcp/k3s based) is another
+option. This incurs management and resource overheads. Future implementation could
+use nested API masters (e.g., kcp/k3s) inside the main Broker cluster for better
+isolation, as long as a management framework is in place to allow automated scaling
+(e.g,. CRD based).
+- Multi-tenancy as the default behavior: `subctl create tenant` also deploys the
+broker (by reconciliation in the axon operator) and the `subctl deploy broker`
+ command is merged into it. This will probably involve further changes and will
+ possibly break some backward compatibility.
+- Tenant management operations can be extracted to a separate CLI tool.
+E.g. `axonctl`.
+
+## Open Questions
+
+- Do we need to update helm too?
+- Do we still need to support the `--namespace` CLI arg? Is it still required to
+support multiple ClusterSets in classic (non-Axon) submariner?
+- How do we authenticate and authorize users beyond the Broker Admin (e.g., k8s
+service accounts)? Should we support additional principles (e.g., users internal
+to a tenant) and permissions within a ClusterSet?
+- Do we create a new repo for axon-cp operator? Or put it in
```

Comment:
IMHO it should be a separate project building on top of Submariner, with it's own tooling and CLI